From: Ben Hutchings <benh@debian.org>
Date: Tue, 30 Nov 2021 23:37:53 +0100
Subject: bpf, tests: Change divide-by-zero test case

This is extracted from commit 21ccaf21497b72f42133182716a42dbf573d314b
"bpf: add further test cases around div/mod and others".

The kselftests for BPF were not added until 4.10, but the test_bpf
module is present and will be used when running kselftests from a
later kernel version.

test_bpf bypasses the verifier and therefore the divide-by-zero checks
that have moved to there.  The divide-by-zero test case will therefore
crash on x86 and some other architectures.  Change it to divide by -1
as done in the above commit.

Cc: Daniel Borkmann <daniel@iogearbox.net>
Cc: Alexei Starovoitov <ast@kernel.org>
Signed-off-by: Ben Hutchings <benh@debian.org>
---
 lib/test_bpf.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

--- a/lib/test_bpf.c
+++ b/lib/test_bpf.c
@@ -1882,10 +1882,14 @@ static struct bpf_test tests[] = {
 		{ { 4, 0 }, { 5, 10 } }
 	},
 	{
-		"INT: DIV by zero",
+		/* This one doesn't go through verifier, but is just raw insn
+		 * as opposed to cBPF tests from here. Thus div by 0 tests are
+		 * done in test_verifier in BPF kselftests.
+		 */
+		"INT: DIV by -1",
 		.u.insns_int = {
 			BPF_ALU64_REG(BPF_MOV, R6, R1),
-			BPF_ALU64_IMM(BPF_MOV, R7, 0),
+			BPF_ALU64_IMM(BPF_MOV, R7, -1),
 			BPF_LD_ABS(BPF_B, 3),
 			BPF_ALU32_REG(BPF_DIV, R0, R7),
 			BPF_EXIT_INSN(),
